---
title: "RNA associations with BCR and TCR repertoire"
author: "Stephen-John Sammut"
output: html_notebook
editor_options: 
  chunk_output_type: console
---


```{r load-data}
rm(list=ls())

library (Biostrings)
library (dplyr)
library (edgeR)
library (fgsea)
library (ggrepel)
library (GSVA)
library (MASS)
library (mclust)
library (mixtools)
library (pheatmap)
library (rstatix)
library (stringr)

# dir.base should point to the location of the main project directory 
dir.base <- "~/BCR-Immunosurveillance/"
source (paste0(dir.base,"R/loadData.R"))

```

```{r calculate-clone-poportions}

# This function loads the Count_per_sample_clusters_merged* files
# generated by MRDARCY which contain the number of UMIs/clone/sample and:

# 1. calculates the proportion of each clone/sample (using duplicate_count)
# 2. classifies clones as stem, clade or private.

# table structure returned:
# BCR clone ID | number of sites clone in | class of clone | Sample ID | proportion | donor

getCloneProportions <- function(file.mrd.CPSCM, donor){
  
  clones <- read.table(file.mrd.CPSCM,stringsAsFactors = F,
                       header = T,comment.char = "",row.names = 1)
  
  # Perform sample overlap analysis in donors that have >1 metastatic site sampled
  if (ncol(clones)>1){
    
    # Compute percentage of UMIs mapped to BCR clones for each sample
    propClones <- data.frame(prop.table(as.matrix(clones),margin = 2)*100)
    # How many metastatic samples is the clone present in?
    numOfSitesWithClone <- apply(propClones, 1, function(x) sum(x>0))
    propClones$numSites <- numOfSitesWithClone
    
    # Classify clones as Stem, Clade or Private
    propClones$class <- "Clade"
    propClones[numOfSitesWithClone==ncol(clones), "class"] <- "Stem"
    propClones[numOfSitesWithClone==1, "class"] <- "Private"
    propClones$class   <- factor(propClones$class,levels=c("Stem","Clade","Private"))
    propClones$cloneID <- rownames(propClones)
    
    p <- reshape2::melt(propClones, id.vars=c("cloneID","numSites","class"))
    # Remove clones with a proportion of 0
    p <- p[p$value!=0,]
    p$donor    <- donor
    p$variable <- gsub("\\.", "-", as.character(p$variable))
    colnames(p)[colnames(p)=="variable"]<-"Sample_ID"
    colnames(p)[colnames(p)=="value"]<-"prop"
    
    p
  }
}

```

```{r load-MrDarcyData}

# This function loads the MrDarcy data files to create a 
# table with structure:
# BCR clone ID | BCR sequence ID | Sample ID | c_call | duplicate_count | SHM count

loadMrDarcyData <- function(file.mrd.MC,file.mrd.fasta, bcrShm){
  
  # Load MRDARCY merge_clustering file to link sequence ID <-> clone ID
  # File structure: Clone ID | BCR sequence ID | Sample ID
  clusters <- read.table(file.mrd.MC, stringsAsFactors = F, header = T,comment.char = "")
  clusters$Sequence_ID <- sapply(strsplit(clusters$Sequence_ID,"\\|"),"[",1)
  clusters <- clusters[!duplicated(clusters),]
  colnames(clusters)[1]<-"cloneID"
  
  # Load MRDARCY FASTA file containing BCR sequences and get number of UMIs
  # (duplicate_count) per isotype, per sequence ID, per sample
  masterFasta <- readDNAStringSet(file.mrd.fasta)
  fastaHeader <- strsplit(names(masterFasta),"\\|")
  stopifnot(length(unique(sapply(fastaHeader,"[",2)))==1)
  fasta <- sapply(strsplit(sapply(fastaHeader,"[",1),"__"),"[",2)
  fasta <- do.call(rbind,strsplit(fasta,"_"))
  class(fasta) <- "numeric"
  fasta <- data.table(fasta,stringsAsFactors = F)
  colnames(fasta)  <- strsplit(unique(sapply(fastaHeader,"[",2)),"_")[[1]]
  fasta$sequenceID <- sapply(strsplit(sapply(fastaHeader,"[",1),"__"),"[",1)
  fasta$Sample_ID     <- sapply(fastaHeader,"[",3)
  fasta$sequence    <- as.character(as.character(masterFasta))
  fasta <-reshape2::melt(fasta,measure.vars=strsplit(unique(sapply(fastaHeader,"[",2)),"_")[[1]])
  fasta <-fasta[fasta$value!=0,]
  fasta <- merge(fasta,bcrShm,by="sequence")
  fasta$sequence <- NULL
  
  fasta$id    <- paste0(fasta$sequenceID,fasta$Sample_ID)
  clusters$id <- paste0(clusters$Sequence_ID,clusters$Sample_ID)
  
  # Create a table that links BCR sequence ID <-> clone ID <-> Sample ID <-> SHM
  clusters <- merge(clusters, fasta[,c(3:6)], by.x="id", by.y="id")
  clusters$id<-NULL
  colnames(clusters)[4]<-"c_call"
  colnames(clusters)[5]<-"duplicate_count"
  data.table(clusters)
  
}

```


```{r calculate-isotype-distributions}

# This function generates a table which summarises the UMI count (duplicate_count)
# per MRDARCY clone, per sample, per BCR isotype

getCloneIsotypeDistributions <- function(x){
  x$mu <- NULL
  # Sum isotype counts by cloneID and sample
  # Return table with structure:
  # Clone ID | Sample ID | counts of IGH isotypes
  igh <- c("IGHA1", "IGHA2", "IGHD", "IGHE", "IGHG1", "IGHG2", "IGHG3", "IGHG4", "IGHM")
  e <- reshape2::dcast(x,cloneID+Sample_ID~c_call, value.var = "duplicate_count",fun.aggregate = sum)
  
  # add IGH isotypes that might not have not been detected in sample
  classNotInCol <- igh[!igh %in% colnames(e)]
  for (i in classNotInCol) e[[i]]<-0
  
  # reorder column names
  ord <- c(1,2,order(colnames(e)[c(3:ncol(e))])+2)
  e <- e[, ord]
  data.table(e)
}


```

```{r calculate-shm}

# This function annotates each sequence with the corresponding MRDARCY clone ID
# and SHM rate
# Return table with structure:
# BCR sequence ID | BCR clone ID | Sample ID | SHM count

getCloneShm <- function(x){
  x <- x[,c(2,1,3,6)]
  x <- x[!duplicated(x),]
  data.table(x)
}

```

```{r load-bcr-data-mets}

# Load metastatic BCR counts data file
# This file can be provided once approval from the Data Access Committee is secured.
bcrData.mets <- readRDS(file$bcr.counts.mets)
bcrData.mets <- rbindlist(bcrData.mets)
bcrData.mets$sequence <- getVDJSequence(bcrData.mets)

# Summarise BCR data: number of mutations (SHM) per BCR sequence
bcrData.mets <- bcrData.mets[, colnames(bcrData.mets) %in% c("sequence","mu_count_total"), with=F]
bcrData.mets <- bcrData.mets[!duplicated(bcrData.mets),]
bcrData.mets <- bcrData.mets[, mean(mu_count_total), by = c("sequence")]
colnames(bcrData.mets)[2] <- "mu"
bcrData.mets$mu <- ceiling(bcrData.mets$mu)

cloneProp.mets.list   <- list()
cloneShmIso.mets.list <- list()
cloneIso.mets.list    <- list()
cloneShm.mets.list    <- list()

# Load metastatic data MRDARCY output
for (s in list.dirs(paste0(dir.mrdarcy.bcr.mets,"output"),recursive = F)) {
  donor <- basename(s)
  cat("Computing summary metrics for donor: ", donor,"\n")
  file.mrd.CPSCM <- paste0(s,"/Count_per_sample_clusters_merged_", donor,".txt.gz")
  file.mrd.MC    <- paste0(s,"/Merge_clustering_", donor,".txt.gz")
  file.mrd.fasta <- paste0(s,"/Combined_fully_reduced_A_sequences_", donor,".fasta.gz")
  
  cloneProp.mets.list[[donor]]    <- getCloneProportions(file.mrd.CPSCM, donor)
  cloneShmIso.mets.list[[donor]]  <- loadMrDarcyData(file.mrd.MC, file.mrd.fasta, bcrData.mets)
  cloneIso.mets.list[[donor]]     <- getCloneIsotypeDistributions(cloneShmIso.mets.list[[donor]])
  cloneShm.mets.list[[donor]]     <- getCloneShm(cloneShmIso.mets.list[[donor]])
}

# perform tidy up and remove large objects from memory
rm(bcrData.mets,s,donor, file.mrd.CPSCM, file.mrd.MC, file.mrd.fasta)
gc(verbose=F)

```


```{r load-bcr-data-early}

# Load early BCR counts data file
# This file can be provided once approval from the Data Access Committee is secured.
bcrData.early <- readRDS(file$bcr.counts.early)
bcrData.early <- rbindlist(bcrData.early)
bcrData.early$sequence <- getVDJSequence(bcrData.early)

# Summarise BCR data: number of mutations (SHM) per BCR sequence
bcrData.early <- bcrData.early[, colnames(bcrData.early) %in% c("sequence","mu_count_total"), with=F]
bcrData.early <- bcrData.early[!duplicated(bcrData.early),]
bcrData.early <- bcrData.early[, mean(mu_count_total), by = c("sequence")]
colnames(bcrData.early)[2] <- "mu"
bcrData.early$mu <- ceiling(bcrData.early$mu)

cloneProp.early.list   <- list()
cloneShmIso.early.list <- list()
cloneIso.early.list    <- list()
cloneShm.early.list    <- list()

# Load MRDARCY output (Early cancer)
for (s in list.dirs(paste0(dir.mrdarcy.bcr.early,"output"),recursive = F)) {
  donor <- basename(s)
  cat("Computing summary metrics for donor: ", donor,"\n")
  file.mrd.CPSCM <- paste0(s,"/Count_per_sample_clusters_merged_", donor,".txt.gz")
  file.mrd.MC    <- paste0(s,"/Merge_clustering_", donor,".txt.gz")
  file.mrd.fasta <- paste0(s,"/Combined_fully_reduced_A_sequences_", donor,".fasta.gz")
  
  cloneProp.early.list[[donor]]    <- getCloneProportions(file.mrd.CPSCM, donor)
  cloneShmIso.early.list[[donor]]  <- loadMrDarcyData(file.mrd.MC, file.mrd.fasta, bcrData.early)
  cloneIso.early.list[[donor]]     <- getCloneIsotypeDistributions(cloneShmIso.early.list[[donor]])
  cloneShm.early.list[[donor]]     <- getCloneShm(cloneShmIso.early.list[[donor]])
}

# perform tidy up and remove large objects from memory
rm(bcrData.early, s, donor, file.mrd.CPSCM, file.mrd.MC, file.mrd.fasta)
rm(getCloneIsotypeDistributions, getCloneProportions, getCloneShm, loadMrDarcyData)
gc(verbose=F)

```

```{r combine-early-metastatic-data}

cloneProp <- rbind(cbind(rbindlist(cloneProp.mets.list), cohort="Metastatic"),
                   cbind(rbindlist(cloneProp.early.list), cohort="Early"))
rm(cloneProp.early.list, cloneProp.mets.list)


cloneShm <- rbind(cbind(rbindlist(cloneShm.mets.list), cohort="Metastatic"),
                  cbind(rbindlist(cloneShm.early.list), cohort="Early"))
rm(cloneShm.early.list, cloneShm.mets.list)


cloneIsotypeProp <- rbind(cbind(rbindlist(cloneIso.mets.list),cohort="Metastatic"),
                          cbind(rbindlist(cloneIso.early.list),cohort="Early"))
rm(cloneIso.early.list, cloneIso.mets.list)


cloneShmIsoProp <- rbind(cbind(rbindlist(cloneShmIso.mets.list),cohort="Metastatic"),
                         cbind(rbindlist(cloneShmIso.early.list),cohort="Early"))
rm(cloneShmIso.early.list, cloneShmIso.mets.list)



```

```{r load-metadata}

metadata.mets$cohort <- "Metastatic"

md.early <- cloneProp[,c("donor","Sample_ID")]
md.early <- md.early[grepl("^T",md.early$donor),]
md.early$Sample<- sapply(strsplit(md.early$Sample_ID,"-"),"[",2)
md.early$cohort <- "Early"
md.early <- md.early[!duplicated(md.early),]
md.early <- md.early[,c(2:4)]
colnames(md.early)<-colnames(metadata.mets)
metadata <- rbind(md.early,metadata.mets)

rm(md.early,metadata.early,metadata.mets,metadata.mets.noAbbrev)
```


```{r load-expression-data}

# Load RNA expression data
expression.early <-  readRDS(file$rna.expr.counts.early)
expression.met   <- readRDS(file$rna.expr.counts.mets)
expression.met   <- expression.met[rownames(expression.met) %in% rownames(expression.early),]
table(rownames(expression.early)==rownames(expression.met))

expression <- cbind(expression.met,expression.early)
expression <- expression[,match(metadata$Sample.ID,colnames(expression))]
stopifnot(sum(colnames(expression)!=metadata$Sample.ID)==0)

# load gene lengths
geneLengths <- data.frame(fread(file$res.ensgToGeneLen))
table((geneLengths$gene)==rownames(expression))

# calculate log2 FPKM
y    <- DGEList(expression,genes = geneLengths)
keep <- filterByExpr(y,group = metadata$cohort)
y <- y[keep, , keep.lib.sizes=FALSE]
y    <- calcNormFactors(y, method = "TMM")

FPKM <- log2(rpkm(y,normalized.lib.sizes = T))

rm(expression.early,expression.met,y)
```


```{r clone-classification-stem-clade-private}

# Classify BCR clones as stem, clade or private depending on whether they are
# observed in all, some, or one metastases from the same donor 

# Calculate mean clone size by donor
cloneProp.mean <- cloneProp[, mean(prop), by = c("cloneID","class","donor","cohort")]
colnames(cloneProp.mean)[5]<-"prop"
cloneProp.mean$prop.log <- log(cloneProp.mean$prop)

stat.test <- cloneProp.mean %>%
  group_by(donor) %>%
  wilcox_test(prop.log~class) %>% 
  add_significance()

stat.test <- stat.test %>% add_xy_position(x = "class")

# Boxplots showing distribution of BCR clone sizes in stem, clade, and private BCR clones across early breast cancer (n= 4487 stem, 4569 clade, 85439 private unique BCR clones) and metastatic cancer (n= 1268 stem, 12022 clade, 142161 private unique BCR clones) cohorts. 
table(cloneProp.mean$class,cloneProp.mean$cohort)
max(stat.test[stat.test$p.adj.signif=="****","p.adj"])
max(stat.test[stat.test$p.adj.signif=="***","p.adj"])
max(stat.test[stat.test$p.adj.signif=="**","p.adj"])


p1 <- ggplot(cloneProp.mean[cloneProp.mean$cohort=="Metastatic",],aes(x=class,y=prop.log))+
  geom_boxplot(outlier.size = 0.5,width=0.8,aes(fill=class))+
  labs(x="",y="Mean BCR clone size/donor (log %UMIs)")+
  scale_fill_manual(values = cols.StemCladePrivate)+
  facet_wrap(~donor,nrow=2)+
  scale_y_continuous(limit=c(-9.2,5))+
  guides(fill="none")+
  stat_pvalue_manual(stat.test[grep("E",stat.test$donor),],bracket.nudge.y = 0.5, hide.ns = TRUE, size = 5)+
  theme_manuscript(base_size = 12)+
  theme(strip.background = element_blank(),
        strip.text = element_text(face="bold"),
        panel.grid.major.x = element_blank())

p2 <- ggplot(cloneProp.mean[cloneProp.mean$cohort=="Early",],aes(x=class,y=prop.log))+
  geom_boxplot(outlier.size = 0.5,width=0.8,aes(fill=class))+
  labs(x="",y="Mean BCR clone size/donor (log %UMIs)")+
  scale_fill_manual(values = cols.StemCladePrivate)+
  facet_wrap(~donor,nrow=2)+
  guides(fill="none")+
  stat_pvalue_manual(stat.test[grep("T",stat.test$donor),],bracket.nudge.y = 0.5, hide.ns = TRUE, size = 5)+
  theme_manuscript(base_size = 12)+
  scale_y_continuous(limit=c(-9.2,5))+
  theme(strip.background = element_blank(),
        strip.text = element_text(face="bold"),
        panel.grid.major.x = element_blank())

pdf(paste0(dir.out.plots,"EFig4a.pdf"),height=12.5/2.54,width=34/2.54)
ggarrange(p2,p1,widths = c(2.3,1))
dev.off()


# On classifying BCR clones as stem, clade or private depending on whether they were observed in all, some, or one metastasis from the same donor, respectively, we observed that immunosurveiling clones were significantly enlarged (stem > clade > private, P<2.2x10-16, ordinal regression, Extended Data Fig. 4a).

d <- polr(factor(cloneProp.mean$class, levels=c("Stem","Clade","Private"), ordered=T) ~ 
            cloneProp.mean$prop.log, Hess=TRUE)
ctable <- coef(summary(d))
pval   <- pnorm(abs(ctable[, "t value"]), lower.tail = FALSE) * 2
pval[1]

rm (d, pval, ctable, stat.test)

```


```{r clone-classification-ABCD}

# Classify clones as:
# A private and expanded
# B shared and expanded
# C private and unexpanded
# D shared and unexpanded (Fig. 3a). 

cloneProp$prop.log <- log(cloneProp$prop)

# Identify expanded/unexpanded cut offs using MClust
set.seed(1985)
m <- Mclust(cloneProp$prop.log,modelNames = "E")
plot(m,what="classification")
plot(m,what="density")

# cut off of 7 chosen to ensure that each class is represented in each sample
# see later
cutOff <- max(m$data[m$classification==7])
par(mfrow=c(1,2))
plot(m,what="density")
title("Mclust density")
abline(v=cutOff,col="red")
plot(m,what="classification")
abline(v=cutOff,col="red")
title("Mclust classes")

cloneProp$freqClass <- "none"
pdf(paste0(dir.out.plots,"MClust-cutoffs.pdf"),width=5, height=5)
par(mfrow=c(2,2))
for (s in unique(cloneProp$Sample_ID)){
  values <- cloneProp[cloneProp$Sample_ID==s,]$prop.log
  plot(density(values),main=paste(s),xlim = c(-10, 1))                
  abline(v=cutOff,col="red")
  cloneProp[cloneProp$Sample_ID==s & cloneProp$prop.log>=cutOff & cloneProp$numSites==1,"freqClass"] <- "A"
  cloneProp[cloneProp$Sample_ID==s & cloneProp$prop.log>=cutOff & cloneProp$numSites>1, "freqClass"] <- "B"
  cloneProp[cloneProp$Sample_ID==s & cloneProp$prop.log<cutOff  & cloneProp$numSites==1,"freqClass"] <- "C"
  cloneProp[cloneProp$Sample_ID==s & cloneProp$prop.log<cutOff  & cloneProp$numSites>1, "freqClass"] <- "D"
}
dev.off()

# check distribution of clones across samples given MClust cut-off
table(cloneProp$freqClass,cloneProp$Sample_ID)


#expanded clones less than 10% of total repertoire
prop.table(table(cloneProp$freqClass))["A"]+
prop.table(table(cloneProp$freqClass))["B"]


# B cell clones present within multiple sites were significantly enlarged compared with site-specific private clones, with BCR clone size correlating with the number of metastatic sites in which these were observed (Fig. 3a, P<2.2x10-16, linear regression),
summary(lm(cloneProp$prop.log~cloneProp$numSites))

# Harmonise clone classes: Clones where clone size is above the expanded cutoff threshold
# in some sites (i.e. expanded) and below the threshold in others (i.e. unexpanded) 
# are classified as expanded.
cloneProp$freqClass <- factor(cloneProp$freqClass,levels=c("C","D","A","B"),ordered = T)

cloneProp$tempID <- paste0(cloneProp$donor,"_",cloneProp$cloneID)

x <- as.data.frame.matrix(table(cloneProp$tempID,cloneProp$freqClass))

# if a clone is in class B and D should be reassigned to B
isExpanded <- rownames(x[x$D!=0 & x$B!=0,])
cloneProp[cloneProp$tempID %in% isExpanded,"freqClass"] <- "B"
# if a clone is in class A and C should be reassigned to A
isExpanded <- rownames(x[x$A!=0 & x$C!=0,])
cloneProp[cloneProp$tempID %in% isExpanded,"freqClass"] <- "A"
cloneProp$tempID <- NULL


# Figure 3a: plot the distribution of mean BCR clone size across the number of metastatic sites detected
cloneProp.mean <- cloneProp[, mean(prop), by = c("cloneID","numSites","donor","cohort")]
colnames(cloneProp.mean)[5]<-"prop"
cloneProp.mean$prop.log <- log(cloneProp.mean$prop)
minUMI <- min(cloneProp.mean$prop.log)
maxUMI <- max(cloneProp.mean$prop.log)

# Plot Figure 3a
pdf(paste0(dir.out.plots,"Fig3a.pdf"),height=9/2.54,width=22/2.54)

ggplot(data = cloneProp.mean, aes(x=as.factor(numSites), y=prop.log))+
  geom_violin(fill=NA,color=NA)+
  annotate("rect", xmin = 0.5, xmax = 1.5, ymin = minUMI, ymax = cutOff, 
           fill="#EEC643", color="white", size=0.6)+
  annotate("rect", xmin = 1.5, xmax = 9.5, ymin = minUMI, ymax = cutOff,
           fill="#79B4A9", color="white", size=0.6)+
  annotate("rect", xmin = 0.5, xmax = 1.5, ymin = cutOff, ymax = maxUMI+0.2, 
           fill="#C06E52", color="white", size=0.6)+
  annotate("rect", xmin = 1.5, xmax = 9.5, ymin = cutOff, ymax = maxUMI+0.2,
           fill="#D5CAD6", color="white", size=0.6)+
  annotate("text",x=0.6, y = maxUMI+0.2,label="A",fontface="bold",color="#C06E52",size=5)+
  annotate("text",x=9.4, y = maxUMI+0.2,label="B",fontface="bold",color="#777077",size=5)+
  annotate("text",x=0.6, y = minUMI,label="C",fontface="bold",color="#BE9E35",size=5)+
  annotate("text",x=9.4, y = minUMI,label="D",fontface="bold",color="#486C65",size=5)+
  geom_violin(aes(fill=(numSites)), scale = "width",trim=T)+
  geom_boxplot(width=0.2,fill="white",outlier.colour = NA)+
  labs(x="Number of sites",y="Mean BCR clone size/donor (log %UMI)")+
  scale_fill_viridis_c()+
  theme_manuscript(base_size = 12)+
  facet_grid(~cohort,scales="free")+
  theme(panel.grid.major.x = element_blank())+
  guides(fill="none")

dev.off()

h=cloneProp.mean
h$numSites<-factor(h$numSites,ordered = T)

mm=polr(formula = numSites~prop.log, data = h, Hess = TRUE)
(ctable <- coef(summary(mm)))
p <- pnorm(abs(ctable[, "t value"]), lower.tail = FALSE) * 2
p



#Violin and box plots showing distribution of mean BCR clone size across sampling time points in early breast cancer (n=94495 unique BCR clones) and across number of sites in metastatic breast cancer (n=155451 unique BCR clones). BCR clones classified as private expanded (class A, n=10507 clones), shared expanded (class B, n=6358 clones), private unexpanded (class C, n=217093 clones), shared unexpanded (class D, n=15988 clones).

k <- cloneProp[,c("cloneID","donor","freqClass","cohort")]
k <- k[!duplicated(k),]
table(k$cohort)
table(k$freqClass)
table(k$freqClass,k$cohort)

rm(k, m, x, cutOff, maxUMI, minUMI,isExpanded, s, p1, p2, values, cloneProp.mean)


```


```{r screen-for-pathogenic-antigens}

# Is there an enrichment of BCR sequences with known binding to viral 
# or bacterial antigens in these four clonal categories?

s    <- cloneShm
s$mu <- NULL
s$id <- paste0(s$cloneID,"_",s$Sample_ID)

j    <- cloneProp
j$id <- paste0(j$cloneID,"_",j$Sample_ID)
j    <- merge(j,s[,c(1,5)],by="id")


# Load results from pathogen antigen screen - Metastatic
# File format: Reference AA sequence | Screen CDR3 sequence | number of mismatches | substring hit
screenRes <- readRDS(paste0(dir.out.data,"Ab-screen-Pathogen.Rdata"))
screenRes.early <- screenRes$early
screenRes.mets  <- screenRes$metastatic

# Load cancer CDR3 reference:
db.cancer <- readRDS(paste0(dir.out.data,"SequenceIDs_CDR3_match.metastatic.RData"))
statsList.mets <- list()
# Iterate by number of CDR3 mismatches
for (diff in c(1:3)){
  r <- screenRes.mets[screenRes.mets$difference<=diff,]
  # Which BCR sequences have a CDR3 match?
  bcrsWithCDR3match <- unique(db.cancer[db.cancer$cdr3_seq %in% r$screen_seq,]$sequence_id)
  n <- j[j$cohort=="Metastatic",]
  n$numMatchBCR   <- ifelse(n$Sequence_ID %in% bcrsWithCDR3match,1,0)
  n$numNoMatchBCR <- ifelse(!n$Sequence_ID %in% bcrsWithCDR3match,1,0)
  n <- data.table(n[,c(5,10,12:13)])
  n <- n[, lapply(.SD, sum, na.rm=TRUE), by=list(Sample_ID,freqClass), .SDcols=c("numMatchBCR", "numNoMatchBCR") ] 
  n$propBCRmatch <- n$numMatchBCR/(n$numMatchBCR+n$numNoMatchBCR)
  n$class <- paste0("<=",diff," mismatches")
  statsList.mets[[as.character(diff)]] <- n
}

# Early breast cancer
db.cancer <- readRDS(paste0(dir.out.data,"SequenceIDs_CDR3_match.early.RData"))

statsList.early <- list()
# Iterate by number of CDR3 mismatches
for (diff in c(1:3)){
  r <- screenRes.early[screenRes.early$difference<=diff,]
  # Which BCR sequences have a CDR3 match?
  bcrsWithCDR3match <- unique(db.cancer[db.cancer$cdr3_seq %in% r$screen_seq,]$sequence_id)
  n <- j[j$cohort=="Early",]
  n$numMatchBCR   <- ifelse(n$Sequence_ID %in% bcrsWithCDR3match,1,0)
  n$numNoMatchBCR <- ifelse(!n$Sequence_ID %in% bcrsWithCDR3match,1,0)
  n <- data.table(n[,c(5,10,12:13)])
  n <- n[, lapply(.SD, sum, na.rm=TRUE), by=list(Sample_ID,freqClass), .SDcols=c("numMatchBCR", "numNoMatchBCR") ] 
  n$propBCRmatch <- n$numMatchBCR/(n$numMatchBCR+n$numNoMatchBCR)
  n$class <- paste0("<=",diff," mismatches")
  statsList.early[[as.character(diff)]] <- n
}


rm(n, r, s, j, db.cancer, screenRes,diff, bcrsWithCDR3match,screenRes.early,screenRes.mets)

# combine datasets
x1 <- cbind(rbindlist(statsList.mets), cohort="Metastatic BC")
x2 <- cbind(rbindlist(statsList.early), cohort="Early BC")
x <- rbind(x1,x2)
x$class <- factor(x$class,levels=c("<=3 mismatches","<=2 mismatches","<=1 mismatches"))
x$freqClass <- factor(x$freqClass,levels=c("C","D","A","B"),ordered = T)


pdf(paste0(dir.out.plots,"EFig4b.pdf"),height=9/2.54,width=14/2.54)
ggplot(x,aes(x=freqClass,y=propBCRmatch,fill=freqClass))+
  geom_boxplot(outlier.size = 0.5)+
  stat_compare_means(label="p.format",method = "anova")+
  labs(x="Class",y="Proportion of BCRs/sample associated\nwith viral and bacterial antigens")+
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))+
  scale_fill_manual(values = cols.cloneClass)+
  facet_grid(cohort~class,scales="free")+
  theme_manuscript(base_size = 12)+
  theme(strip.background = element_blank())+
  guides(fill="none")
dev.off()


#Boxplots showing proportion of unique BCRs/sample with CDR3 sequences matching a reference antibody database with known binding to viral or bacterial antigens with ≤3, ≤2 and ≤1 CDR3 mismatches, across the four BCR clone classes (n=23 metastatic cancer samples, n=25 early breast cancer samples, P values calculated using two-sided analysis of variance (ANOVA)).
table(x$freqClass,x$class,x$cohort)

rm(x1,x2,x,statsList.early,statsList.mets)

```


```{r which-clone-classes-have-highest-proportion-in-organs}

# Quantify relative proportion of expanded BCRs grouped by site

x <- data.table(cloneProp[,c("Sample_ID","prop","donor","freqClass","cohort")])
x <- x[, lapply(.SD,sum), by=.(Sample_ID,freqClass,donor,cohort)]
x <- merge(x,metadata[,c("Sample.ID","Disease.site"),],by.x="Sample_ID",by.y="Sample.ID") 

sitesToKeep<-c("Liver","Lung/pleura","Lymph node","diagnosis","midway","surgical")

x <- x[x$Disease.site%in% sitesToKeep,]
table(x$freqClass, x$Disease.site)
table(x$donor)

# Expanded immunosurveilling clones (clone class B) comprised the majority of tumour infiltrating B cell clones (Fig. 3b) and were present at higher proportions in the liver and lung/pleura compared to private expanded clones (class A)

plot.fig3b <- ggplot(x, aes(x=freqClass,y=prop,fill=freqClass))+
  geom_boxplot(width=0.8,outlier.size = 0.5)+
  geom_point(size=0.8)+
  labs(x="Clone class",y="%BCR UMIs/sample")+
  facet_wrap(cohort~Disease.site,nrow=2)+
  stat_compare_means(comparisons = list(c(3,4)),label.y.npc = 0.7)+
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))+
  theme_manuscript(base_size = 12)+
  scale_fill_manual(values=cols.cloneClass)+
  guides(fill="none")+
  theme(strip.background = element_blank(),
        panel.grid.major.x = element_blank())

pdf(paste0(dir.out.plots,"Fig3b.pdf"),height=12/2.54,width=10/2.54)
plot.fig3b
dev.off()


rm(x, plot.fig3b,sitesToKeep)
```

```{r pgen-analysis}

# Is there a difference in CDR3 pgen in these four clonal categories?

# load OLGA output
pgen.all <- readRDS(paste0(dir.out.data,"CDR3-OLGA-pgen.RData"))

# remove CDR3s with probability of 0
pgen <- data.frame(rbind(pgen.all$early,pgen.all$metastatic))
pgen <- pgen[pgen$pgen!=0,]
pgen$pgen <- signif(pgen$pgen,8)
pgen <- pgen[!duplicated(pgen),]

# check that there are no duplicates
t <- table(pgen$cdr3)
t[t>1]

# metastatic cohort - associate CDR3 with sequence ID
imgt <- fread(paste0(dir.mrdarcy.bcr.mets,"IMGT/output/5_AA-sequences.txt.gz"),fill=T,sep = "\t")
imgt <- imgt[,c("Sequence ID","JUNCTION")]
imgt <- merge(imgt,pgen,by.x="JUNCTION",by.y="cdr3")

x <- cloneShm[cloneShm$cohort=="Metastatic",]
x$mu <- NULL
x$ID <- paste0(x$cloneID,"_",x$Sample_ID)

# Merge clone proportion statistics with sequence IDs
j    <- cloneProp[cloneProp$cohort=="Metastatic",]
j$ID <- paste0(j$cloneID,"_",j$Sample_ID)
j <- j[,c(9,10)]
j <- merge(j,x,by="ID")
j <- merge(j,imgt,by.x="Sequence_ID",by.y="Sequence ID")
j$donor <- substring(j$Sample_ID,1,4)

head(j)
j <- j[,c("freqClass","JUNCTION","pgen","donor")]

# keep unique CDR3s and remove any in multiple classes
j <- j[!duplicated(j),]
e <- dcast(j,JUNCTION+donor~freqClass,value.var="pgen")
f <- apply(e[,c(3:6)],1,function(x) sum(is.na(x)))
e <- e[f==3,]
keep <- paste0(e$donor,e$JUNCTION)
j <- j[paste0(j$donor,j$JUNCTION) %in% keep,]

j.mets <- j
j.mets$cohort <- "Metastatic"


# early cancer cohort - associate CDR3 with sequence ID
imgt <- fread(paste0(dir.mrdarcy.bcr.early,"IMGT/output/5_AA-sequences.txt.gz"),fill=T,sep = "\t")
imgt <- imgt[,c("Sequence ID","JUNCTION")]
imgt <- merge(imgt,pgen,by.x="JUNCTION",by.y="cdr3")

x <- cloneShm[cloneShm$cohort=="Early",]
x$mu <- NULL
x$ID <- paste0(x$cloneID,"_",x$Sample_ID)

# Merge clone proportion statistics with sequence IDs
j    <- cloneProp[cloneProp$cohort=="Early",]
j$ID <- paste0(j$cloneID,"_",j$Sample_ID)
j <- j[,c(9,10)]
j <- merge(j,x,by="ID")
j <- merge(j,imgt,by.x="Sequence_ID",by.y="Sequence ID")
j$donor <- substring(j$Sample_ID,1,4)

# keep unique CDR3s and remove any in multiple classes
j <- j[,c("freqClass","JUNCTION","pgen","donor")]
j <- j[!duplicated(j),]

e <- dcast(j,JUNCTION+donor~freqClass,value.var="pgen")
f <- apply(e[,c(3:6)],1,function(x) sum(is.na(x)))
e <- e[f==3,]
keep <- paste0(e$donor,e$JUNCTION)
j <- j[paste0(j$donor,j$JUNCTION) %in% keep,]

j.early <- j
j.early$cohort <- "Early"

j <- rbind(j.mets,j.early)
j$freqClass <- factor(j$freqClass,levels=c("C","D","A","B"))

median_pgen.early <- j[j$cohort=="Early",] %>%
  group_by(freqClass) %>%
  summarize(median=median(log10(pgen)))

median_pgen.mets <- j[j$cohort=="Metastatic",] %>%
  group_by(freqClass) %>%
  summarize(median=median(log10(pgen)))


Fig3b.1 <- ggplot(j[j$cohort=="Early",],aes(group=freqClass,x=log10(pgen)))+
  geom_density(aes(color=freqClass),bw=1.5,size=0.8)+
  geom_vline(data = median_pgen.early, 
             aes(xintercept = median,color = freqClass),
             linetype="dashed")+ 
  coord_cartesian(xlim=c(-47,-5))+
  scale_y_continuous(limits=c(0,0.08),breaks = c(0,0.02,0.04,0.06,0.08))+
  labs(y="Probability density",x=expression(log["10"](P["gen"])))+
  scale_color_manual(name="Clone class", values = cols.cloneClass)+
  scale_fill_manual(name="Clone class", values = cols.cloneClass)+
  theme_manuscript(base_size = 12)+
  facet_grid(cohort~.)+
  guides(fill=guide_legend(nrow=2),color=guide_legend(nrow=2))


Fig3b.2 <- ggplot(j[j$cohort=="Metastatic",],aes(group=freqClass,x=log10(pgen)))+
  geom_density(aes(color=freqClass),bw=1.5,size=0.8)+
  geom_vline(data = median_pgen.mets, 
             aes(xintercept = median,color = freqClass),
             linetype="dashed")+ 
  coord_cartesian(xlim=c(-47,-5))+
  scale_y_continuous(limits=c(0,0.08),breaks = c(0,0.02,0.04,0.06,0.08))+
  labs(y="Probability density",x=expression(log["10"](P["gen"])))+
  scale_color_manual(name="Clone class", values = cols.cloneClass)+
  scale_fill_manual(name="Clone class", values = cols.cloneClass)+
  theme_manuscript(base_size = 12)+
  facet_grid(cohort~.)+
  guides(fill=guide_legend(nrow=2),color=guide_legend(nrow=2))

ggarrange(Fig3b.1,Fig3b.2,nrow=2,align = "hv")

# Data from donors with more than one tumour site sampled shown (early breast cancer cohort: all donors (n=10), metastatic breast cancer cohort donors: 308, 315, 323, 330) used.
table(j$donor)

# Healthy donor PBMC analysis

pgen.pbmc <- pgen.all$pbmc_healthy
pgen.pbmc$class <- paste0(pgen.pbmc$IGH, pgen.pbmc$SHM)
pgen.pbmc$class <- ifelse(pgen.pbmc$class=="UnswitchedUnmutated","Antigen inexperienced","Antigen experienced")

head(pgen.pbmc)

#further filtering, keep cdr3s in one class only
e <- dcast(pgen.pbmc,JUNCTION+donor_id~class,value.var="pgen")
f <- apply(e[,c(3:4)],1,function(x) sum(is.na(x)))
e <- e[f==1,]
keep <- paste0(e$donor,e$JUNCTION)
pgen.pbmc <- pgen.pbmc[paste0(pgen.pbmc$donor,pgen.pbmc$JUNCTION) %in% keep,]

median_pgen <- pgen.pbmc %>%
  group_by(class) %>%
  summarize(median=median(log10(pgen)))

pgen.pbmc$cohort<-"PBMC"

Fig3b.3 <- 
  ggplot(pgen.pbmc,aes(x=log10(pgen),group=class,color=class))+
  geom_density(size=0.8)+
  geom_vline(data = median_pgen, 
             aes(xintercept = median,color = class),
             linetype="dashed")+
  # geom_vline(xintercept = cutOff)+
  coord_cartesian(xlim = c(-47,-5))+
  labs(y="Probability density",x=expression(log["10"](P["gen"])))+
  scale_color_manual(name="", values = c("blue","red"))+
  theme_manuscript(base_size = 12)+
  facet_grid(cohort~.)+
  guides(fill=guide_legend(nrow=2),color=guide_legend(nrow=2))


pdf(paste0(dir.out.plots,"Fig3c.pdf"),height=20/2.54,width=8/2.54)
ggarrange(Fig3b.1,Fig3b.2,Fig3b.3,nrow = 3,align = "hv")
dev.off()

rm(Fig3b.1,Fig3b.2,Fig3b.3,pgen,pgen.pbmc,median_pgen,median_pgen.early,median_pgen.mets,f,keep,t)
rm(j.early,j.mets,j,e,imgt,pgen.all,x)
```


```{r isotypes-metastatic-vs-early}

I <- cloneIsotypeProp
I$cloneID <- NULL
I <- data.frame(I[, lapply(.SD,sum), by=.(Sample_ID,cohort)])

# Calculate IGH proportion/sample
I[,grepl("IGH",colnames(I))] <- prop.table(as.matrix(I[,grepl("IGH",colnames(I))]),
                                           margin = 1)*100
I$unswitched <- I$IGHD+I$IGHM

pdf(paste0(dir.out.plots,"EFig4c.pdf"),height=7/2.54,width=5/2.54)
ggplot(I,aes(x=cohort,y=(unswitched),fill=cohort))+
  geom_boxplot(outlier.size = 0.5,width=0.6)+
  labs(x="Disease stage",y="Unswitched BCRs/sample(%)")+
  theme_manuscript(base_size = 12)+
  scale_fill_manual(values=c("#6baed6","#cb181d"))+
  stat_compare_means(label = "p.format")+
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))+
  theme(panel.grid.major.x = element_blank())+
  guides(fill="none")
dev.off()

#Data from all donors used (early breast cancer cohort: n=10 donors, 25 samples; metastatic breast cancer cohort n=8 donors, 27 samples) used.
t=table(substring(I$Sample_ID,1,4))
sum(t[grepl("E",names(t))])
sum(t[grepl("T",names(t))])

rm(I)
```


```{r isotype-distribution-cloneClasses}

# Determine %BCRs unswitched across 4 clone classes

temp        <- cloneProp[,c("cloneID","Sample_ID","freqClass")]
temp$tempID <- paste0(temp$cloneID,"_",temp$Sample_ID)
temp        <- temp[,c("tempID","freqClass")]

I <- cloneIsotypeProp
I$tempID <- paste0(I$cloneID,"_",I$Sample_ID)
I <- merge(I,temp,by="tempID")
I[,tempID:=NULL]


# Summarise isotype counts by BCR clone class
I$cloneID <- NULL
I <- I[, lapply(.SD,sum), by=.(Sample_ID,freqClass,cohort)]
P <- data.frame(merge(I, metadata[,c("Sample.ID","Disease.site")],
                      by.x="Sample_ID",by.y="Sample.ID"),stringsAsFactors = F)

# Calculate IGH proportion/sample
P[,grepl("IGH",colnames(P))] <- prop.table(as.matrix(P[,grepl("IGH",colnames(P))]),
                                           margin = 1)*100
P$unswitched <- P$IGHD+P$IGHM

plot.Fig3d <- ggplot(P,aes(x=freqClass,y=(unswitched),fill=freqClass))+
  geom_boxplot(outlier.size = 0.5,width=0.6)+
  labs(x="Clone class",y="Unswitched BCRs/sample(%)")+
  stat_compare_means(comparisons = list(c("A","B"),c("C","B")),label.y.npc = 0.7)+
  theme_manuscript(base_size = 12)+
  facet_grid(~cohort)+
  theme(panel.grid.major.x = element_blank())+
  scale_fill_manual(values=cols.cloneClass)+
  guides(fill="none")
plot.Fig3d

pdf(paste0(dir.out.plots,"Fig3d.pdf"),height=6.2/2.54,width=10/2.54)
plot.Fig3d
dev.off()

table(substring(P$Sample_ID,1,4))

# 4 clone classes
table(P$Sample_ID)

# 48 samples
length(unique(P$Sample_ID))

#48*4 = 192 sample/clone class combinations
nrow(P)


# compute significance between classes
for (j in LETTERS[1:4]){
  cat(j,":", wilcox.test(P[P$cohort=="Metastatic" &P$freqClass==j,"unswitched"],
                         P[P$cohort!="Metastatic" &P$freqClass==j,"unswitched"])$p.value,"\n")
}

table(P$freqClass)
rm (I, P, plot.Fig3d, j)


# isotype proportion distribution in early and late disease
K <- cloneIsotypeProp
K$cloneID <- NULL
K <- K[, lapply(.SD,sum), by=.(Sample_ID,cohort)]
K <- data.frame(K)
K[,c(3:11)]<-prop.table(as.matrix(K[,c(3:11)]),margin = 1)
K <- melt(K,id.vars=c("Sample_ID","cohort"))
K$time <- sapply(strsplit(K$Sample_ID,"-"),"[",2)
K$time[!K$time %in% c("diagnosis","midway","surgical")]<-"metastatic"
K$time <- factor(K$time,levels=c("diagnosis","midway","surgical","metastatic"),
                 labels=c("Pre","Mid","Post","Metastatic"))

plot.Fig3f <- ggplot(K[!K$variable %in% c("IGHD","IGHE","IGHG4"),],aes(x=time,y=value*100,fill=time))+
  geom_boxplot(width=0.8,outlier.size = 0.5)+
  labs(x="Disease stage",y="Isotype proportion (%)")+
  theme_manuscript(base_size = 12)+
  scale_fill_manual(values=c("#08519c","#6baed6", "#c6dbef","#cb181d"))+
  guides(fill="none")+
  theme(panel.grid.major.x = element_blank(),
        strip.background = element_blank())+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))+
  facet_wrap(~variable)
plot.Fig3f


pdf(paste0(dir.out.plots,"Fig4f.pdf"),height=11.2/2.54,width=15/2.54)
plot.Fig3f
dev.off()

length(table(K$Sample_ID))
unique(substring(K$Sample_ID,1,4))

#obtain monotonic p values for early disease
o <- K[K$cohort=="Early",]
o$time<-factor(o$time,levels=c("Pre","Mid","Post"),ordered = T)
pval <- character()
for (ig in unique(as.character(K$variable))){
  mm <- polr(formula = time~value, data = o[o$variable==ig,], Hess = TRUE)
  ctable <- coef(summary(mm))
  p <- pnorm(abs(ctable[, "t value"]), lower.tail = FALSE) * 2
  pval=rbind(pval,cbind("early",ig,p[1]))
}

# Wilcox metastatic vs early
o<-K
for (ig in unique(as.character(K$variable))){
  p=wilcox.test(K[K$cohort=="Early" & K$variable==ig,4],K[K$cohort!="Early" & K$variable==ig,4])$p.value
  pval=rbind(pval,cbind("late",ig,p[1]))
}

pval = data.frame(pval)
pval

#=================
temp        <- cloneProp[,c("cloneID","Sample_ID","freqClass")]
temp$tempID <- paste0(temp$cloneID,"_",temp$Sample_ID)
temp        <- temp[,c("tempID","freqClass")]
I <- cloneIsotypeProp
I$tempID <- paste0(I$cloneID,"_",I$Sample_ID)
I <- merge(I,temp,by="tempID")
I[,tempID:=NULL]

class<-list()
for (type in c(LETTERS[1:4])){
  K<-I
  K<-K[K$freqClass==type,]
  K$cloneID<-NULL
  K$freqClass=NULL
  K<-K[, lapply(.SD,sum), by=.(Sample_ID,cohort)]
  K=data.frame(K)
  K[,c(3:11)]<-prop.table(as.matrix(K[,c(3:11)]),margin = 1)
  K<-melt(K,id.vars=c("Sample_ID","cohort"))
  K$time<-sapply(strsplit(K$Sample_ID,"-"),"[",2)
  K$time[!K$time %in% c("diagnosis","midway","surgical")]<-"metastatic"
  K$time <- factor(K$time,levels=c("diagnosis","midway","surgical","metastatic"),
                   labels=c("Pre","Mid","Post","Metastatic"))
  K$class <-type
  class[[type]]<-K
}
class<-do.call(rbind,class)

class <- class[!class$variable %in% c("IGHD","IGHE","IGHG3","IGHG4"),]

class$class<-factor(class$class,levels=c("C","D","A","B"))
plot.eFig4g <- ggplot(class,aes(x=time,y=value*100,fill=time))+
  geom_boxplot(width=0.8,outlier.size = 0.5)+
  labs(x="Disease stage",y="BCR isotype proportion (%)")+
  theme_manuscript(base_size = 12)+
  scale_fill_manual(values=c("#08519c","#6baed6", "#c6dbef","#cb181d"))+
  guides(fill="none")+
  theme(panel.grid.major.x = element_blank(),
        strip.background = element_blank())+
  scale_y_continuous(limits = c(0,105),expand = expansion(mult = c(0, 0.1)), breaks=c(0,25,50,75,100))+
  facet_grid(variable~class)

pdf(paste0(dir.out.plots,"EFig4g.pdf"),height=19/2.54,width=16.5/2.54)
plot.eFig4g
dev.off()

table(substring(class$Sample_ID,1,4))

stats <- character()
for (ig in unique(as.character(class$variable))){
  for (clone in c("C","D","A","B")){

  p=wilcox.test(class[class$cohort=="Early" & class$class==clone & class$variable==ig,4],
                class[class$cohort!="Early" & class$class==clone & class$variable==ig,4])$p.value
  stats=rbind(stats,c(ig,clone,p))
  }
}
stats=data.frame(stats)
stats[,3]<-as.numeric(stats[,3])
stats$p.adj<-p.adjust(stats[,3],method="BH")
stats$p.adj2=signif(stats$p.adj,2)
stats


rm(K, o, p, ig, mm, plot.Fig3f, ctable)


# Associate with expression
geneList <- readRDS(file$res.msigdb.c5)
geneList <- geneList[grepl("GOBP_ISOTYPE_SWITCHING_TO_IGG_ISOTYPES|GOBP_ISOTYPE_SWITCHING_TO_IGA_ISOTYPES",
                           names(geneList),ignore.case = T)]
m <- t(gsva(as.matrix(FPKM),geneList,method="ssgsea"))
m <- melt(m)
m <- merge(m,metadata,by=1)

#n=52 samples
unique(as.character(m$Var1))


m$Var2<-factor(m$Var2,levels =c("GOBP_ISOTYPE_SWITCHING_TO_IGA_ISOTYPES","GOBP_ISOTYPE_SWITCHING_TO_IGG_ISOTYPES"),
               labels=c("Isotype switching to\nIGA subtypes\n(GO:0048290)","Isotype switching to\nIGG subtypes\n(GO:0048291)"))

pdf(paste0(dir.out.plots,"EFig4h.pdf"),height=8/2.54,width=9/2.54)

ggplot(m,aes(x= cohort,y=value,fill=cohort))+
  geom_boxplot(outlier.size = 0.5,width=0.6)+
  facet_wrap(~Var2)+
  labs(x="",y="ssGSEA score")+
  scale_fill_manual(values=c("#6baed6","#cb181d"))+
  theme_manuscript(base_size = 12)+
  stat_compare_means(label="p.format")+
  guides(fill="none")+
  theme(strip.background = element_blank(),
        panel.grid.major.x = element_blank())
dev.off()

rm(geneList,m, I, temp)

```


```{r shm-distribution-metastatic-vs-early}

# Calculate mean SHM per clone
cloneShmProp <- data.table(data.frame(cloneShm))
cloneShmProp[,Sequence_ID:=NULL]
cloneShmProp <- cloneShmProp[, lapply(.SD,mean), by=.(cloneID,Sample_ID,cohort)]

unique((cloneShmProp$Sample_ID))

pdf(paste0(dir.out.plots,"EFig4e.pdf"),height=7/2.54,width=5/2.54)
ggplot(cloneShmProp,aes(x=cohort,y=mu, fill=cohort))+
  geom_boxplot(outlier.size = 0.5,width=0.6)+
  labs(x="Disease stage",y="Mean mutation count/clone/sample")+
  scale_fill_manual(values=c("#6baed6","#cb181d"))+
  stat_compare_means(label = "p.format")+
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))+
  theme_manuscript(base_size = 12)+
  theme(panel.grid.major.x = element_blank())+
  guides(fill="none")
dev.off()
```


```{r AICDA-expression}

# Load previous RNA expression data
# calculate FPKM
y    <- DGEList(expression,genes = geneLengths)
#keep lowly expressed genes
keep <- rowSums(cpm(y)>0.1) >= 10
y <- y[keep, , keep.lib.sizes=FALSE]
y    <- calcNormFactors(y, method = "TMM")
expr <- (rpkm(y,normalized.lib.sizes = T))
expr <- data.frame(t(expr[rownames(expr) %in% c("ENSG00000111732"),,drop=F]))
head(expr)
colnames(expr)<-c("AICDA")
expr <- merge(expr,metadata,by.x=0,by.y=1)

#n=52 samples
expr$Row.names

pdf(paste0(dir.out.plots,"EFig4f.pdf"),height=7/2.54,width=5.5/2.54)

ggplot(expr,aes(x=cohort,y=AICDA,fill=cohort))+
  geom_boxplot(outlier.size = 0.5,width=0.6)+
  labs(x="Disease stage",y="Expression (FPKM)")+
  stat_compare_means(label = "p.format")+
  theme_manuscript(base_size = 12)+
  scale_fill_manual(values=c("#6baed6","#cb181d"))+
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))+
  theme(panel.grid.major.x = element_blank(),
        strip.background = element_blank())+
  guides(fill="none")
dev.off()
```


```{r classify-shm-mixed-model}

# classify each BCR into one of four SHM categories
set.seed(1985)
cloneShmIsoProp$ID <- paste0(cloneShmIsoProp$cloneID,cloneShmIsoProp$Sample_ID)

cloneProp <- data.frame(cloneProp)
temp    <- cloneProp[,colnames(cloneProp) %in% c("cloneID","Sample_ID","freqClass")]
temp$ID <- paste0(temp$cloneID,temp$Sample_ID)

x<- cloneShmIsoProp
x$ID <- NULL
x$donor <- substring(x$Sample_ID,1,4)

plot_mix_comps <- function(x, mu, sigma, lam) {
  lam * dnorm(x, mu, sigma)
}

mat <- x
mutations_log <- log10(as.numeric(mat$mu)+1)
mutations <- as.numeric(mat$mu)

mixmdl <- normalmixEM(mutations_log, k = 2)

bimodal_peaks <- (10^(mixmdl$mu))-1
bimodal_upper_interval <- (10^(mixmdl$mu+(mixmdl$sigma*1)))-1
bimodal_lower_interval <- (10^(mixmdl$mu-(mixmdl$sigma*1)))-1

data.frame(x = mixmdl$x) %>%
  ggplot() +
  geom_histogram(aes(x, ..density..), binwidth = 0.05, colour = "black", 
                 fill = "white") +
  stat_function(geom = "line", fun = plot_mix_comps,
                args = list(mixmdl$mu[1], mixmdl$sigma[1], lam = mixmdl$lambda[1]),
                colour = "red", lwd = 1.5) +
  stat_function(geom = "line", fun = plot_mix_comps,
                args = list(mixmdl$mu[2], mixmdl$sigma[2], lam = mixmdl$lambda[2]),
                colour = "blue", lwd = 1.5) +
  ylab("Density") + 
  xlab("log10(mutations+1)")

dev.off()
h = hist(mutations,max(mutations)-min(mutations),plot = F)

summary(mutations)
plot.EFig4d_1 <- ggplot(data.frame(mutations),aes(x=mutations))+
  geom_density(size=0.5,fill="gray90")+
  geom_vline(xintercept = bimodal_upper_interval[1],col="red",linetype="dashed")+
  geom_vline(xintercept = bimodal_lower_interval[2],col="darkgreen",linetype="dashed")+
  geom_vline(xintercept = bimodal_upper_interval[2],col="blue",linetype="dashed")+
  labs(x="Number of mutations",y="Density")+
  scale_x_continuous(limits=c(-4,77))+ 
  scale_y_continuous(limits=c(0,0.08),breaks=c(0,0.02,0.04,0.06,0.08))+
  theme_manuscript()



thresholds = sort(c(0, bimodal_upper_interval, bimodal_lower_interval[2], max(mutations)+1))
mutational_type = rep("UNKNOWN", length(mutations))
types = c("No SHM", "Low SHM","High SHM", "Very high SHM")

for (i in c(2:length(thresholds))){
  w = intersect(which(mutations>=thresholds[i-1]), which(mutations<thresholds[i]))
  mutational_type[w] = types[i-1]
}

table(mutational_type)

mat$mutational_type=mutational_type
mat$ID <- paste0(mat$cloneID,mat$Sample_ID)

cloneProp <- data.frame(cloneProp)
temp    <- cloneProp[,colnames(cloneProp) %in% c("cloneID","Sample_ID","freqClass")]
temp$ID <- paste0(temp$cloneID,temp$Sample_ID)

mat <- merge(mat,temp[,c("ID","freqClass")],by="ID",all=T)
head(mat)
mat_full <- mat
```

```{r classify-shm-byclass}

mat <- mat[!is.na(mat$freqClass),]
classes = c("A","B","C","D")
samples = unique(sort(mat$Sample_ID))

#Number of BCRs mutational type/class/Sample
t = table(mat$Sample_ID, mat$mutational_type,mat$freqClass)
t

list_proportions_per_type = list()

for(c in c(1:length(classes))){
  m = t[,,classes[c]]
  # normalise
  m = prop.table(m,margin = 1)*100
  list_proportions_per_type[[c]] <- m
}
names(list_proportions_per_type) = classes

st<-list()
for(c in c(1:length(types))){
  groups <- list()
  for (j in c(1:length(classes))){
    groups[[j]] <-  list_proportions_per_type[[classes[j]]][,types[c]]
  }
  names(groups) = classes
  h <- melt(do.call(rbind,groups))
  h$class<-types[c]
  st[[c]] <-  h
}


h <- do.call(rbind,st)  
h$Var1=factor(h$Var1,levels=c("C","D","A","B"))
h$class <- factor(h$class,levels=c("No SHM","Low SHM","High SHM","Very high SHM"))

h$cohort<-ifelse(grepl("E",as.character(h$Var2)),"Early","Metastatic")

#48 samples
unique(h$Var2)

plot.EFig4d_2 <- ggplot(h,aes(x=Var1,y=value,fill=Var1))+
  geom_boxplot(width=0.8,outlier.size = 0.5)+
  labs(x="Clone class",y="%BCRs/SHM class/\nclone class/sample")+
  
  facet_grid(cohort~class)+
  stat_compare_means(label="p.format",method = "kruskal.test")+
  theme_manuscript(base_size = 12)+
  scale_fill_manual(values=cols.cloneClass)+
  guides(fill="none")+
  theme(panel.grid.major.x = element_blank(),
        strip.background = element_blank())+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))

pdf(paste0(dir.out.plots,"EFig4d.pdf"),height=16/2.54,width=16/2.54)

ggarrange(plot.EFig4d_1+
            theme(plot.margin = unit(c(0.5,1,0.5,0.5), "cm")),
          plot.EFig4d_2,nrow=2,heights = c(1.2,2), align = "h")

dev.off()
```


```{r classify-shm-bydisease}

#Number of BCRs mutational type/class/Sample
t = table(mat_full$Sample_ID, mat_full$mutational_type)

m <- prop.table(t,margin = 1)*100
m <- merge(
  data.frame(`No/Low SHM`=m[,"Low SHM"]+m[,"No SHM"]),
  data.frame(`High/Very high SHM`=m[,"High SHM"]+m[,"Very high SHM"]),by=0)


g=reshape2::melt(m)
colnames(g)[1]<-"Var1"
colnames(g)[2]<-"mutational_type"

g$cohort<-ifelse(grepl("E",g$Var1),"Metastatic","Early")
g$time<-sapply(strsplit(as.character(g$Var1),"-"),"[",2)

g$time[!g$time %in% c("diagnosis","midway","surgical")]<-"Metastatic"
g$mutational_type <- factor(g$mutational_type,
                            levels=c("No.Low.SHM","High.Very.high.SHM"),
                            labels=c("No/low SHM","High/Very high SHM"))

g$donor<-sapply(strsplit(as.character(g$Var1),"-"),"[",1)

g$time<-factor(g$time,levels=c("diagnosis","midway","surgical","Metastatic"),ordered = T)

unique(g$Var1)
pdf(paste0(dir.out.plots,"Fig3e.pdf"),height=5.5/2.54,width=6/2.54)
ggplot(g[g$mutational_type=="High/Very high SHM",],aes(x=time,y=value,fill=time))+
  geom_boxplot(width=0.8,outlier.size = 0.5)+
  labs(x="Disease stage",y="Highly mutated BCRs/sample (%)")+
  theme_manuscript(base_size = 12)+
  scale_fill_manual(values=c("#08519c","#6baed6", "#c6dbef","#cb181d"))+
  guides(fill="none")+
  theme(panel.grid.major.x = element_blank(),
        strip.background = element_blank())+
  scale_y_continuous(expand = expansion(mult = c(0, 0.1)))

dev.off()

q <- g[g$mutational_type=="High/Very high SHM" & g$cohort!="Metastatic",]
mm=polr(formula = time~value, data = q, Hess = TRUE)
(ctable <- coef(summary(mm)))
p <- pnorm(abs(ctable[, "t value"]), lower.tail = FALSE) * 2
p

q <- g[g$mutational_type=="High/Very high SHM",]
wilcox.test(q[q$cohort=="Metastatic","value"], q[q$cohort!="Metastatic","value"])

```



```{r calculate-proportion-of-B-clones}

propOfB <- data.table(cloneProp)
propOfB <- propOfB[propOfB$freqClass=="B",]
propOfB <- propOfB[, .(propB=sum(prop)), by=.(Sample_ID)] 
propOfB$donor <- substr(propOfB$Sample_ID,1,4)
propOfB <- merge(propOfB,metadata,by.x="Sample_ID", by.y="Sample.ID")
propOfB <- data.frame(propOfB)

```


```{r calculate-csr-shm-class}

# 1. Calculate % unswitched BCRs
I <- cloneIsotypeProp
I$cloneID <- NULL
I <- I[, lapply(.SD,sum), by=.(Sample_ID,cohort)]
iso <- data.frame(merge(I, metadata[,c("Sample.ID","Disease.site")], 
                        by.x="Sample_ID", by.y="Sample.ID"), stringsAsFactors = F)
totCounts <- rowSums(iso[,grepl("IGH",colnames(iso))])
iso[,grepl("IGH",colnames(iso))] <- iso[,grepl("IGH",colnames(iso))]/totCounts*100
iso$unswitched <- iso$IGHD+iso$IGHM


# 2. Calculate mean SHM per sample
shm <- cloneShm
shm$Sequence_ID <- NULL
shm <- shm[, lapply(.SD,mean), by=.(cloneID,Sample_ID,cohort)]
shm$cloneID<-NULL
shm <- shm[, .(mutations=mean(mu)), by=.(Sample_ID)] 
shm <- data.frame(merge(shm, metadata[,c("Sample.ID","Disease.site")], 
                        by.x="Sample_ID", by.y="Sample.ID"), stringsAsFactors = F)

x       <- merge(shm,iso[,c(1,2,13)],by=1)
x$donor <-substring(x$Sample_ID,1,4)


x$class.unswitched <- ifelse(x$unswitched > median(x$unswitched), "Low CSR", "High CSR")
x$class.shm <- ifelse(x$mutations > median(x$mutations), "High SHM", "Low SHM")

head(x)
dim(x)

# is there an association between High CSR+SHM and prop of B?
e <- merge(x,propOfB[,c(1,2)],by=1)
e$fc <- paste0(e$class.unswitched,e$class.shm)
e <- e[!e$fc %in% c("High CSRLow SHM","Low CSRHigh SHM"),]
e$fc <- factor(e$fc,levels=c("Low CSRLow SHM","High CSRHigh SHM"),
               labels=c("Low","High"))
p1 <-ggplot(e,aes(x=fc,y=propB,fill=fc))+
  geom_boxplot(width=0.7,outlier.size = 0.5)+
  geom_point(size=0.5)+
  scale_fill_manual(values = rev(c("#FB6542","#375E97")))+
  labs(x="CSR+SHM class",y="BCR class B isotype proportion (%)")+
  stat_compare_means(label="p.format")+
  theme_manuscript()+
  guides(fill="none")+
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))+
  theme(strip.background = element_blank(),
        panel.grid.major.x = element_blank())

unique(e$donor)
# is there an association between High CSR+SHM and TME?

x <- x[match(colnames(expression),x$Sample_ID),]
table(colnames(expression)==x$Sample_ID)

dictionary <- data.frame(fread(file$res.ensgToHugo),stringsAsFactors = F)

# Calculate T cell inflamed and IFN gamma scores
genes.ifng     <- c("IFNG", "STAT1", "IDO1", "CXCL10", "CXCL9", "HLA-DRA")
genes.inflamed <- c("TIGIT","CD27","CD8A","PDCD1LG2","LAG3","CD274","CXCR6",
                    "CMKLR1","NKG7","CCL5","PSMB10","IDO1","CXCL9","HLA-DQA1",
                    "CD276","STAT1","HLA-DRB1","HLA-E")

genes.ifng     <- dictionary[dictionary$Hugo %in% genes.ifng,"Ensembl.ID"]
genes.inflamed <- dictionary[dictionary$Hugo %in% genes.inflamed,"Ensembl.ID"]

gene.tls.hallmark <- c("CCL19", "CCL21", "CXCL13", "CCR7", "CXCR5", "SELL" , "LAMP3")
gene.tls.hallmark     <- dictionary[dictionary$Hugo %in% gene.tls.hallmark,"Ensembl.ID"]

geneList <- list(`IFN-g score` = genes.ifng, `Inflamed score`=genes.inflamed,TLS=gene.tls.hallmark)

d <- readRDS(file$res.danaher)
d <- d[d$Cell.Type %in% c("B-cells","T-cells"),]
for (h in unique(d$Cell.Type)){
  geneList[[h]]<-d[d$Cell.Type==h,"Gene.stable.ID"]
}

e <- t(gsva(as.matrix(FPKM),geneList,method="gsva",min.sz=1))
e <- merge(e,x,by.x=0,by.y=1)
e <- reshape2::melt(e,id.vars=c("Row.names","mutations","Disease.site",
                                "unswitched","donor","class.unswitched",
                                "class.shm","cohort"))
e$fc <- paste0(e$class.unswitched,e$class.shm)
e <- e[!e$fc %in% c("High CSRLow SHM","Low CSRHigh SHM"),]
e$fc <- factor(e$fc,levels=c("Low CSRLow SHM","High CSRHigh SHM"),
               labels=c("Low","High"))


e$variable<-factor(e$variable,levels=c("B-cells","T-cells","TLS","DC","IFN-g score","Inflamed score"))
p2<- ggplot(e,aes(x= fc,y=value,fill=fc))+
  geom_boxplot(width=0.7,outlier.size = 0.5)+
  geom_point(size=0.5)+
  scale_fill_manual(values = rev(c("#DC493A","#08519C")))+
  labs(x="CSR+SHM class",y="GSVA score")+
  stat_compare_means(label="p.format")+
  theme_manuscript()+
  guides(fill="none")+
  facet_wrap(~variable,nrow=1)+
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))+
  theme(strip.background = element_blank(),
        panel.grid.major.x = element_blank())
p2

pdf(paste0(dir.out.plots,"Fig3h.pdf"),height=5.5/2.54,width=20/2.54)
ggarrange(p1,p2,widths=c(0.25,1))
dev.off()

```

```{r GTEX-part-1, eval=F}
# do this analysis in GTEX

# For GTEx analysis:

# 1. Download GTEx data (1.5Gb) and metadata from: 
# https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz
# https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt

# 2. Update the location gtexDir variable:
#the following files should be present in this directory:
# - GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz
# - GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt.gz
gtexDir <- "~/Dropbox/Stephen/Work/Research/datasets/external/GTEX/v8/RNA/"

#Precomputed matrices are available in the output/data folder, see final section
# if fails follow this (Edit .Renviron and set to 100gb):
# https://stackoverflow.com/questions/51295402/r-on-macos-error-vector-memory-exhausted-limit-reached

file.expr.gtex    <- paste0(gtexDir,"GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz")
file.metadata.gtex <- paste0(gtexDir,"GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt.gz")


# Load GTEx expression and metadata files
expr.gtex.all <- fread(file.expr.gtex)
metadata.gtex <- fread(file.metadata.gtex)[,c("SAMPID","SMTS")]

# Subset GTEx data, retaining only samples obtained from sites that have also
# been sequenced within this study
metadata.gtex <- metadata.gtex[metadata.gtex$SMTS %in%
                                 c("Brain","Breast","Liver","Lung"),]

toKeep <- colnames(expr.gtex.all) %in% 
                             c("Name","Description",metadata.gtex$SAMPID)

expr.gtex.all     <- expr.gtex.all[,toKeep,with=F]
saveRDS(expr.gtex.all,"~/B-cell_immunosurveillance_breast_cancer/output/data/GTEX-summary.RData")

expr.gtex.all<-readRDS("~/B-cell_immunosurveillance_breast_cancer/output/data/GTEX-summary.RData")

# retain IGH only
expr.gtex.iso <- expr.gtex.all[grepl("^IGHA1$|^IGHA2$|^IGHG1$|^IGHG2$|^IGHG3$|^IGHG4$|^IGHD$|^IGHM$|^IGHE$",
                            expr.gtex.all$Description),]
expr.gtex.iso <- data.frame(expr.gtex.iso,row.names = 2)
expr.gtex.iso$Name<-NULL

gc(verbose=F)

totTpm <- colSums(expr.gtex.iso)
unswitched <- apply(expr.gtex.iso[rownames(expr.gtex.iso) %in% c("IGHD","IGHM"),],2,sum)
unswitched <- unswitched/totTpm
plot(density(unswitched))
unswitched<-unswitched
#50th percentile BCR CSR/sample
class.unswitched <- ifelse(unswitched > quantile(unswitched)[3], 
                                "High", "Low")
class.unswitched <- factor(class.unswitched, 
                                levels=rev(c("High","Low")))

expr.gtex.all$Description <- NULL
expr.gtex.all$Name <- sapply(strsplit(expr.gtex.all$Name,"\\."),"[",1)
expr.gtex.all <- expr.gtex.all[!duplicated(expr.gtex.all$Name),]
expr.gtex.all <- data.frame(expr.gtex.all,row.names = 1)
expr.gtex.all <- as.matrix(expr.gtex.all)
expr.gtex.all <- expr.gtex.all[apply(expr.gtex.all,1, function(x) sum(x)>100),]
dim(expr.gtex.all)
tpmExpression <- (log2(expr.gtex.all+0.1))
dictionary <- data.frame(fread(file$res.ensgToHugo),stringsAsFactors = F)

# Calculate cytolytic activity score
genes.cyt <- c("GZMA", "PRF1")
genes.cyt <- dictionary[dictionary$Hugo %in% genes.cyt, "Ensembl.ID"]

# Calculate T cell inflamed and IFN gamma scores
genes.ifng     <- c("IFNG", "STAT1", "IDO1", "CXCL10", "CXCL9", "HLA-DRA")
genes.inflamed <- c("TIGIT","CD27","CD8A","PDCD1LG2","LAG3","CD274","CXCR6",
                    "CMKLR1","NKG7","CCL5","PSMB10","IDO1","CXCL9","HLA-DQA1",
                    "CD276","STAT1","HLA-DRB1","HLA-E")
genes.tls.hallmark <- c("CCL19", "CCL21", "CXCL13", "CCR7", "CXCR5", "SELL" , "LAMP3")

genes.ifng     <- dictionary[dictionary$Hugo %in% genes.ifng,"Ensembl.ID"]
genes.inflamed <- dictionary[dictionary$Hugo %in% genes.inflamed,"Ensembl.ID"]
genes.tls.hallmark     <- dictionary[dictionary$Hugo %in% genes.tls.hallmark,"Ensembl.ID"]

#do a gsva 
geneList <- list(`CYT score`=genes.cyt, `IFN-g score` = genes.ifng, `Inflamed score`=genes.inflamed, TLS=genes.tls.hallmark)

d <- readRDS(file[["res.danaher"]])
d <- d[d$Cell.Type %in% c("B-cells","T-cells","Cytotoxic cells","NK cells"),]

for (h in unique(d$Cell.Type)){
  geneList[[h]]<-d[d$Cell.Type==h,"Gene.stable.ID"]
}


# As this is very compute heavy, split across three randomised batches
sam  <- sample(c(1:ncol(tpmExpression)),ncol(tpmExpression), replace = F)
sam1 <- sam[1:1000]
sam2 <- sam[1001:2000]
sam4 <- sam[2001:3000]
sam3 <- sam[3001:ncol(tpmExpression)]

e1 <- t(gsva((tpmExpression[,sam1]),geneList,method="gsva",min.sz > 1))
e2 <- t(gsva((tpmExpression[,sam2]),geneList,method="gsva",min.sz > 1))
e3 <- t(gsva((tpmExpression[,sam3]),geneList,method="gsva",min.sz > 1))
e4 <- t(gsva((tpmExpression[,sam4]),geneList,method="gsva",min.sz > 1))

e <- rbind(e1,e2,e3,e4)
e <- merge(e,data.frame(class.unswitched),by.x=0,by.y=0)
dim(e)
rm(sam,sam1,sam2,sam3,e1,e2,e3)

saveRDS(e,paste0(dir.out.data,"GTEx-CSR-GSVA.RData"))
```

```{r GTEX-part-2}

e <- readRDS(paste0(dir.out.data,"GTEx-CSR-GSVA.RData"))
dim(e)

e <- reshape2::melt(e,id.vars=c("Row.names","class.unswitched"))
e=e[e$variable %in% c("B-cells","T-cells","TLS","IFN-g score","Inflamed score"),]
e$variable<-factor(e$variable,levels=c("B-cells","T-cells","TLS","IFN-g score","Inflamed score"))

pdf(paste0(dir.out.plots,"EFig4i.pdf"),height=9.5/2.54,width=10/2.54)

ggplot(e,aes(x= class.unswitched,y=value,fill=class.unswitched))+
  geom_boxplot(width=0.7,outlier.size = 0.5)+
  scale_fill_manual(values = rev(c("#FB6542","#375E97")))+
  labs(x="Proportion of unswitched transcripts (GTEx)",y="GSVA score")+
  stat_compare_means(label = "p.format")+
  theme_manuscript()+
  guides(fill="none")+
  facet_wrap(~variable,nrow=2)+
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.1)))+
  theme(strip.background = element_blank(),
        panel.grid.major.x = element_blank())

dev.off()

p.adjust(c(0.086,7.5e-11,0,1.9e-09,0,0.00048,1.3e-6,9.7e-8))
```


